### DISCUSSION TITLE
WooCommerce and WordPress private APIs: should we start using them?

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#51879] [WooCommerce, Blocks, and Site Editing]
Traditionally, we avoid using “experimental WordPress APIs” in WooCommerce. Recently, during a discussion on this topic started from [this issue](https://github.com/woocommerce/woocommerce/issues/47924), we identified two potential solutions: implementing custom controls independently or relying on “experimental WordPress APIs.”

Firstly, utilizing experimental APIs allows us to provide valuable feedback to WordPress Core, which can help shape the final version of the APIs. This feedback ensures they meet the community’s needs and address potential issues early in the development process. Furthermore, relying on experimental APIs can reduce the amount of custom code we need to maintain, leading to a cleaner and more maintainable codebase. 

[This is the PR](https://github.com/woocommerce/woocommerce/pull/50729/files) that fixed #47924 via experimental APIs.

Conversely, the primary disadvantage is that experimental APIs are inherently unstable. They may contain bugs or undergo significant changes before becoming stable. To mitigate these risks, we could use experimental APIs only for “atomic” features, such as specific block settings. These features could be conditionally displayed only if the experimental API is available. Additionally, implementing E2E tests can enhance the safety of these implementations, ensuring that any issues are quickly identified and addressed. I want to highlight that the WordPress Core team advises avoiding these APIs:

![image](https://github.com/user-attachments/assets/9a6f7fd8-2f2b-4c95-8922-04c03c19e505)

_https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L87-L95_

By carefully integrating experimental APIs and rigorously testing our implementations, we can take advantage of new WordPress features while maintaining the stability and reliability WooCommerce is known for.
<hr />

This topic may be controversial and present some challenges. After discussing it with @mdperez86 and @danielwrobert, it seems beneficial to make it public to gather feedback.

What are your thoughts on this? If you disagree with this approach, what alternative methods could help facilitate feedback on experimental APIs and ensure their stabilization for being use in WooCommerce?

**Comment by tyxla** [comment]
As a member of Team Calypso (the maintainers of the `@wordpress/components` package), I'd encourage y'all to use private APIs but to actually take it a step further. Utilizing the private API mechanism is a great way to get a necessary feature or component earlier in Gutenberg and allow for some time until it gets stable enough to become a public API. A downside of using private APIs is that they're prone to substantial changes, so using them requires the consumer to be up to date with the changes and actively monitor them. I'd still say it's worth it, because it allows us to keep the product as close to core as possible and follow the same consistent direction for our UIs and interfaces.

**Reply by gigitux** [reply]
Thanks for your feedback!
> I'd encourage y'all to use private APIs but to actually take it a step further. Utilizing the private API mechanism is a great way to get a necessary feature or component earlier in Gutenberg and allow for some time until it gets stable enough to become a public API

Furthermore, as consumers, WooCommerce can help WordPress Core giving some feedback.

As @louwie17 noticed:

> This PR calls __dangerousOptInToUnstableAPIsOnlyForCoreModules for the wordpress/block-editor package. Which I believe will only work if the block-editor package it-self is not instantiated. As it will fail if the above function is already called for the mentioned package see: https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L103-L109
> I would be in favour if we can extend [this package list](https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L12) to include some of our WooCommerce packages.
Meaning we alter this condition: The second argument must be a known @wordpress package that hasn't yet opted into @wordpress/private-apis

If the goal is to encourage plugins to use this function (accepting the associated risks), we should refine both the logic and the messaging surrounding the `__dangerousOptInToUnstableAPIsOnlyForCoreModules` function.

**Reply by tyxla** [reply]
I'd say, hopefully, some of these APIs can be stabilized, and I'd love it if folks from WooCommerce product can help with pushing that forward.

**Reply by jsnajdr** [reply]
Using experimental APIs by plugins, outside core, is going to be controversial. Core is trying quite hard to prevent you from doing that:
1. A package needs to opt-in with a `__dangerousOptIn...` call, and needs to pass an argument where it identifies itself (like `optIn( '@wordpress/fields' )`. There's a hardcoded allowlist of package names, and there are checks about double opt-in with the same name -- you can't "impersonate" a core package, it will be detected. If we add `@woocommerce/*` entries to that allowlist, there will be pushback: why exactly should these packages be privileged and get a hardcoded exception in Core?
2. The `__dangerousOptIn` call is designed to break regularly, with each Core release. The call must supply a confirmation message as an argument (`__dangerousOptIn( 'I understand that...' )`) and that message changes. In PRs like WordPress/gutenberg#62635. In Gutenberg internal usages the opt-in calls are updated at the same time, but if you have a plugin that opts in with message A and you upgrade Core from 6.6 to 6.7, the plugin will break: Core expects message B now! This is all done to further discourage plugins from using the private APIs.

The current policy is that plugins shouldn't use private API at all. And if a plugin can and does use a private API, then it's no longer private, it's effectively public, it just has a weird obfuscated name.

**Reply by jsnajdr** [reply]
Another plugin that tries to use private APIs, with mixed success, is MailPoet. Here you can see how they do it:

https://github.com/mailpoet/mailpoet/blob/trunk/mailpoet/assets/js/src/email-editor/lock-unlock.js

They impersonate as the `@wordpress/edit-post` package, exploiting the fact that this package/script is loaded only on the `post.php` page. And in [this PR](https://github.com/mailpoet/mailpoet/pull/5851) @lysyjan also had to update the confirmation message to the WP 6.6 version. That breaks compatibility with WP 6.5, and that's not great.

**Reply by tyxla** [reply]
What @jsnajdr is saying is true and fair FWIW - thanks for pointing it out. 

Using a private API in a **production** plugin is **not recommended**. 

However, taking the time to use it in a **dev version** (under a feature flag or another limitation mechanism to keep it away from production) makes sense and can allow for simultaneous development in Gutenberg / core and usage in the plugin itself.

**Reply by jsnajdr** [reply]
> However, taking the time to use it in a dev version makes sense

If we want to establish such a policy, where we distinguish between production and dev plugins, it should be somehow supported by the `__dangerouslyOptIn` API. The currently implemented policy is that only Gutenberg can use the private APIs internally, it's completely closed to plugins.

Any ideas how that could look like? And how we tell production plugins apart from dev ones? Would they have to run in debug mode, with something like `SCRIPT_DEBUG` enabled? Or would be have a policy that plugins that use the private APIs cannot be published in the plugins directory?

**Reply by tyxla** [reply]
`SCRIPT_DEBUG` is the first thing that comes to mind. However, I think it will be confusing: a site developer might enable `SCRIPT_DEBUG` to debug their site and they'll be exposed to Woo UI that they don't expect. 

When I talk about dev version, I have something else in mind: literally a dev version that either:
- Lives in a dev or feature branch, and end users never get exposed to it
- Is under a feature flag or experimental toggle, similar to the Gutenberg experiments, but within the boundaries of the plugin or plugin family. 
- Uses any custom mechanism for concealing experimental features that the plugin devs find useful.

**Comment by louwie17** [comment]
Thanks for kick starting this discussion @gigitux 💯 

I am in favour of using the experimental API's for similar reasons as @tyxla stated above.
I think it would be worth having a directory of some kind where we keep track of all the places we use the experimental API. We may also want to add some additional requirements for merging these API's like:
- Should include an E2E test for latest WP version and against latest GB version
- Have unit test coverage ( although that one feels less beneficial as this would be tested with our installed dependencies and not the latest ).

> [This is the PR](https://github.com/woocommerce/woocommerce/pull/50729/files) that fixed https://github.com/woocommerce/woocommerce/issues/47924 via experimental APIs.

This PR calls `__dangerousOptInToUnstableAPIsOnlyForCoreModules` for the `wordpress/block-editor` package. Which I believe will only work if the `block-editor` package it-self is not instantiated. As it will fail if the above function is already called for the mentioned package see: https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L103-L109

I would be in favour if we can extend [this package list](https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L12) to include some of our WooCommerce packages.
Meaning we alter this condition: `The second argument must be a known @wordpress package that hasn't yet opted into @wordpress/private-apis`

**Reply by louwie17** [reply]
> I would be in favour if we can extend [this package list](https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/implementation.js#L12) to include some of our WooCommerce packages.
Meaning we alter this condition: The second argument must be a known @wordpress package that hasn't yet opted into @wordpress/private-apis

Although having said this, this may not be necessary. We could also mimic the `private-apis` package on our end for WooCommerce packages specifically. Since all it does is wrap the private API's behind a `__private` key and the unlock is not bound to a specific package as deemed by this unit test: https://github.com/WordPress/gutenberg/blob/4186b74e7784bf96191231c371b9ef49945d62db/packages/private-apis/src/test/index.js#L117-L132

**Reply by louwie17** [reply]
Hmm I stand corrected, they use `Symbol` as the key and `Symbol("foo") !== Symbol("foo")`

**Reply by gigitux** [reply]
Thanks for your feedback! 

> I think it would be worth having a directory of some kind where we keep track of all the places we use the experimental API. We may also want to add some additional requirements for merging these API's like:
>Should include an E2E test for latest WP version and against latest GB version
Have unit test coverage ( although that one feels less beneficial as this would be tested with our installed dependencies and not the latest ).


💯 Agreed. If we get consensus, I think that the next step is creating documentation and common guidelines about how to introduce and maintain experimental APIs in WooCommerce.

**Comment by Aljullu** [comment]
+1 from my side on prioritizing the use of experimental WordPress APIs over building our custom solutions.

At the same time, I think that before using an experimental API, we should consider a couple of things:

* Every experimental API carries the risk of breaking with future updates. While automated testing can help mitigate this risk, it won't fully eliminate it. So I think we should be wary about using experimental APIs in parts of the codebase we are not actively working on.
* While using experimental APIs might solve a specific blocker for us, the true value lies in contributing upstream, whether by submitting fixes (if something is broken) or providing positive feedback (if everything is working well). I think we should be mindful that these contributions take time.

So I would advocate for being transparent with design/product/leadership about the features that require the usage of experimental APIs, and clearly outlining the pros and cons. Using an experimental API requires some engineering work in the future until the API is stabilized, so if we can't commit some resources to maintain it and contribute upstream, we might want to reevaluate whether that feature should be a priority or it can be put on hold until we have more capacity.

**Comment by costasovo** [comment]
Also +1 on prioritizing using experimental WordPress APIs over building our custom solutions.

In addition, I would like to share my thoughts on this. 

I'm from MailPoet, and we use a couple of private APIs. 

WordPress often introduces cool new stuff we would love to use in our product, but the feature or API is private. I believe that the core team wants to get feedback on the private APIs, and there are plugin developers who want to use these APIs. As I see it, the current policy, "plugins shouldn't use private API at all," puts us in a kind of lose-lose situation where the core team gets feedback only from internal usage, and external/plugin devs must create custom solutions or take the risk of using the private API and be ready to fix it when it suddenly changes.

What makes it worse is that in the WordPress environment, we don't control when we get the changes as they come with every new WordPress or Woo release.

One approach we tried in MailPoet was bundling the WordPress packages with the plugin as we do with other external libraries. That way, we can update the packages at our own pace. This works, but I would discourage using this approach because it increases the size of the JS loaded in the browser. There is also a risk of conflicts with JS loaded from the core.

The current policy seems to have two extremes: if an API is public, we need to maintain BC for ages, or it is private, and third-party developers shouldn't touch it. It would be nice to have something in between. For example, an unstable API would mean it can be marked as deprecated, and consumers need to switch to a newer version, which could be another unstable or stable API. When an API is marked as deprecated, it will be removed in the next release. The deprecated version would coexist with the non-deprecated version for some time. This would give third-party developers more space to try those APIs and react to changes. Perhaps the removal of the deprecated APIs could be somewhat automated.

**Comment by ralucaStan** [comment]
I think we need to align with the Gutenberg way of defining `experimental` and `private`:
The contribution guides explain the differences: [Legacy experimental APIs, plugin-only APIs, and private APIs](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#legacy-experimental-apis-plugin-only-apis-and-private-apis).

@gigitux you mention `This is the PR that fixed #47924 via experimental APIs`. But the PR actually uses private APIs.



According to the contribution guides, experimental APIs are legacy, and they have a deprecation period. New APIs are no longer introduced as experimental. 

So, is your question: WooCommerce and WordPress ~experimental~ *private* APIs: should we start using them?

**Reply by gigitux** [reply]
I was referring to the `private` APIs. I updated the title discussion! Thanks for noticing it! 🙇

**Comment by ralucaStan** [comment]
I also wonder if we need to green-light private APIs usage or if we should have a separate discussion with the maintainers for each API we plan on using. From the outside, it's hard to evaluate the maturity of an API just by how long it's been out there. 

> [This is the PR](https://github.com/woocommerce/woocommerce/pull/50729/files) that fixed https://github.com/woocommerce/woocommerce/issues/47924 via experimental APIs.

@mdperez86 was the private API straightforward, is there anything missing from it that would help us achieve our use case?

**Comment by jsnajdr** [comment]
I opened a Gutenberg issue because this is something that needs to be solved in Gutenberg. The current public/private API system simply forbids to expose anything experimental to plugins. This will need to change, if we want plugins to participate in creating and evolving new APIs, instead of being mere consumers of something that is already finished.

https://github.com/WordPress/gutenberg/issues/66197

**Reply by gigitux** [reply]
Thanks for kicking-off the discussion on Gutenberg. I will start to follow the discussion!

**Comment by gigitux** [comment]
Thank you for your valuable feedback on this discussion. Since a new discussion has started in the Gutenberg repository, this thread will now be closed: https://github.com/WordPress/gutenberg/issues/66197.
